c2rust-analyze/src/rewrite/expr/mir_op.rs

```diff
@@ -485,51 +502,170 @@ impl<'a, 'tcx> ExprRewriteVisitor<'a, 'tcx> {
+                    let _rw = match opt_mutbl {
+                        // Currently a no-op, since `Slice` and `OffsetPtr` are identical.
+                        Some(_mutbl) => (),
+                        None => break,
+                    };
+                    // self.emit(rw);
+                    from.qty = Quantity::OffsetPtr;
+                }
+                (Quantity::OffsetPtr, Quantity::Slice) => {
+                    let _rw = match opt_mutbl {
+                        // Currently a no-op, since `Slice` and `OffsetPtr` are identical.
+                        Some(_mutbl) => (),
+                        None => break,
+                    };
+                    // self.emit(rw);
+                    from.qty = Quantity::Slice;
+                }
+                // `Slice` and `OffsetPtr` convert to `Single` the same way.
+                // TODO: may be better to use `slice.as_ptr()` to avoid panic on 0-length inputs
+                (_, Quantity::Single) => {
+                    let rw = match opt_mutbl {
+                        Some(mutbl) => RewriteKind::SliceFirst { mutbl },
+                        None => break,
+                    };
+                    self.emit(rw);
+                    from.qty = Quantity::Single;
+                }
+
+                // Unsupported cases
+                (Quantity::Single, _) => break,
+                (_, Quantity::Array) => break,
 
-        if from.qty == to.qty && (from.own, to.own) == (Ownership::Mut, Ownership::Imm) {
-            self.emit(RewriteKind::MutToImm);
-            from.own = to.own;
+                // Remaining cases are impossible, since `from.qty != to.qty`.
+                (Quantity::Slice, Quantity::Slice) | (Quantity::OffsetPtr, Quantity::OffsetPtr) => {
+                    unreachable!()
+                }
+            }
         }
 
-        if (from.qty, to.qty) == (Quantity::Single, Quantity::Single)
-            && (from.own, to.own) == (Ownership::Mut, Ownership::Cell)
-        {
-            self.emit(RewriteKind::CellFromMut);
-            from.own = to.own;
+        // Late `Ownership` casts.
+        while from.own != to.own {
+            match self.cast_ownership_one_step(from, to, false) {
+                Some(new_own) => {
+                    from.own = new_own;
+                }
+                None => break,
+            }
         }
 
         if from != to {
-            panic!(
+            error!(
                 "unsupported cast kind: {:?} -> {:?} (original input: {:?})",
                 from, to, orig_from
             );
         }
     }
 
+    fn cast_ownership_one_step(
+        &mut self,
+        from: TypeDesc,
+        to: TypeDesc,
+        early: bool,
+    ) -> Option<Ownership> {
+        match from.own {
+            Ownership::Box => match to.own {
+                Ownership::Raw | Ownership::Imm => {
+                    error!("NYI: cast Box to Imm");
```

Comment:
By the way, for these `NYI`s in `error!`s, can we use `TODO` instead so that we can grep for `TODO`.  Normally these would be `todo!`s, right?

Reply:
I thought the idea with #854 is that we should now use `error!` instead of `panic!`/`todo!` when possible so that the error can be made non-fatal by adjusting the `RUST_LOG_PANIC` env var.

Reply:
Yeah, that's right.  I just mean doing something like `error!("TODO: ...")` so we can easily search for TODOs.  Or we could add a wrapper `todo!` macro in `mod log`.

Reply:
Ah, that makes sense.  I'll do that.